confignode/src/main/java/org/apache/iotdb/confignode/manager/TriggerManager.java

```diff
@@ -132,8 +146,79 @@ public TGetTriggerJarResp getTriggerJar(TGetTriggerJarReq req) {
     }
   }
 
-  public TSStatus transferTrigger(List<TDataNodeLocation> newUnknownDataList) {
-    // TODO implement
+  /**
+   * Step1: Mark Stateful Triggers on UnknownDataNodes as {@link TTriggerState#TRANSFERRING}.
+   *
+   * <p>Step2: Get all Transferring Triggers marked in Step1.
+   *
+   * <p>Step3: For each trigger get in Step2, find the DataNode with the lowest load, then transfer
+   * the Stateful Trigger to it and update this information on all DataNodes.
+   *
+   * <p>Step4: Update the newest location on ConfigNodes.
+   *
+   * @param dataNodeLocationMap The DataNodes with {@link
+   *     org.apache.iotdb.commons.cluster.NodeStatus#Running} State
+   * @return result of transferTrigger
+   */
+  public TSStatus transferTrigger(
+      List<TDataNodeLocation> newUnknownDataList,
+      Map<Integer, TDataNodeLocation> dataNodeLocationMap) {
+    TSStatus transferResult;
+    triggerInfo.acquireTriggerTableLock();
+    try {
+      ConsensusManager consensusManager = configManager.getConsensusManager();
+      NodeManager nodeManager = configManager.getNodeManager();
+
+      transferResult =
+          consensusManager
+              .write(new UpdateTriggersOnTransferNodesPlan(newUnknownDataList))
+              .getStatus();
+      if (transferResult.getCode() != TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
+        return transferResult;
+      }
+
+      List<String> transferringTriggers =
+          ((TransferringTriggersResp)
+                  consensusManager.read(new GetTransferringTriggersPlan()).getDataset())
+              .getTransferringTriggers();
+
+      for (String trigger : transferringTriggers) {
+        TDataNodeLocation newDataNodeLocation =
+            nodeManager.getLowestLoadDataNode(dataNodeLocationMap.keySet());
+
+        transferResult =
+            RpcUtils.squashResponseStatusList(
+                updateTriggerLocation(trigger, newDataNodeLocation, dataNodeLocationMap));
+        if (transferResult.getCode() != TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
+          return transferResult;
+        }
+
+        transferResult =
+            consensusManager
+                .write(new UpdateTriggerLocationPlan(trigger, newDataNodeLocation))
+                .getStatus();
+        if (transferResult.getCode() != TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
+          return transferResult;
+        }
+      }
+    } finally {
+      triggerInfo.releaseTriggerTableLock();
+    }
+
     return new TSStatus(TSStatusCode.SUCCESS_STATUS.getStatusCode());
   }
+
+  public List<TSStatus> updateTriggerLocation(
+      String triggerName,
+      TDataNodeLocation dataNodeLocation,
```

Comment:
parameter `dataNodeLocation` and `dataNodeLocationMap` are similar, maybe we can change `dataNodeLocation` to `targetDataNode`?